File: src/core/intc.zig
//!
//! moestation - PlayStation 2 emulator written in Zig.
//! Copyright (c) 2022 Michelle-Marie Schiller
//!
//! intc.zig - Interrupt controller
//!

const std = @import("std");

const assert = std.debug.assert;

const err  = std.log.err;
const info = std.log.info;

const iop = @import("iop.zig");

/// Interrupt sources
const IntSource = enum(u4) {
    Gs,
    Sbus,
    VblankStart,
    VblankEnd,
    Vif0,
    Vif1,
    Vu0,
    Vu1,
    Ipu,
    Timer0,
    Timer1,
    Timer2,
    Timer3,
    Sfifo,
    Vu0Watchdog,
};

pub const IntSourceIop = enum(u5) {
    VblankStart,
    Gpu,
    Cdvd,
    Dma,
    Timer0,
    Timer1,
    Timer2,
    Sio0,
    Sio1,
    Spu2,
    Pio,
    VblankEnd,
    Dvd,
    Pcmcia,
    Timer3,
    Timer4,
    Timer5,
    Sio2,
    Htr0,
    Htr1,
    Htr2,
    Htr3,
    Usb,
    Extr,
    Fwre,
    Fdma,
};

// INTC registers
var intcMask: u15 = undefined;
var intcStat: u15 = undefined;

// INTC registers (IOP)
var iStat: u25  = undefined;
var iMask: u25  = undefined;
var iCtrl: bool = undefined;

/// Returns I_CTRL
pub fn getCtrl() u32 {
    return @as(u32, @bitCast(u1, iCtrl));
}

/// Returns INTC_MASK
pub fn getMask() u32 {
    return @as(u32, intcMask);
}

/// Returns I_MASK
pub fn getMaskIop(comptime T: type, offset: u2) T {
    var data: T = undefined;

    switch (T) {
        u8  => assert(false),
        u16 => {
            if ((offset & 1) == 0) {
                data = @truncate(u16, iMask);
            } else {
                data = @truncate(u16, iMask >> 16);
            }
        },
        u32 => data = @as(u32, iMask),
        else => unreachable,
    }

    return data;
}

/// Returns INTC_STAT
pub fn getStat() u32 {
    return @as(u32, intcStat);
}

/// Returns I_STAT
pub fn getStatIop() u32 {
    return @as(u32, iStat);
}

/// Sets I_CTRL
pub fn setCtrl(data: u32) void {
    iCtrl = (data & 1) != 0;

    checkInterrupt();
}

/// Sets INTC_MASK, checks for interrupt
pub fn setMask(data: u32) void {
    intcMask = @truncate(u15, data);

    checkInterrupt();
}

/// Sets I_MASK, checks for interrupt
pub fn setMaskIop(comptime T: type, data: T, offset: u2) void {
    switch (T) {
        u8  => assert(false),
        u16 => {
            if ((offset & 1) == 0) {
                iMask = (iMask & 0x1F_0000) | @as(u25, data);
            } else {
                iMask = (@as(u25, data) << 16) | (iMask & 0xFFFF);
            }
        },
        u32 => iMask = @truncate(u25, data),
        else => unreachable,
    }

    checkInterruptIop();
}

/// Sets INTC_STAT, checks for interrupt
pub fn setStat(data: u32) void {
    intcStat &= ~@truncate(u15, data);

    checkInterrupt();
}

/// Sets I_STAT, checks for interrupt
pub fn setStatIop(comptime T: type, data: T, offset: u2) void {
    switch (T) {
        u8  => assert(false),
        u16 => {
            if ((offset & 1) == 0) {
                iStat &= ~@as(u25, data & 0xFFFF);
            } else {
                iStat &= ~(@as(u25, (data & 0xFFFF)) << 16);
            }
        },
        u32 => iStat = ~@truncate(u25, data),
        else => unreachable,
    }

    checkInterruptIop();
}

/// Sends an IOP interrupt request
pub fn sendInterruptIop(src: IntSourceIop) void {
    info("   [INTC (IOP)] {s} interrupt request.", .{@tagName(src)});

    iStat |= @as(u25, 1) << @enumToInt(src);

    checkInterruptIop();
}

fn checkInterrupt() void {
    if ((intcStat & intcMask) != 0) {
        err("  [INTC      ] Unhandled EE interrupt.", .{});

        assert(false);
    }
}

fn checkInterruptIop() void {
    info("   [INTC (IOP)] I_CTRL = {}, I_STAT = 0b{b:0>25}, I_MASK = 0b{b:0>25}", .{iCtrl, iStat, iMask});

    iop.setIntPending(iCtrl and ((iStat & iMask) != 0));
}

File: src/core/intc.cpp
/*
 * moestation is a WIP PlayStation 2 emulator.
 * Copyright (C) 2022-2023  Lady Starbreeze (Michelle-Marie Schiller)
 */

#include "intc.hpp"

#include <cassert>
#include <cstdio>

#include "iop/cop0.hpp"

namespace ps2::intc {

/* IOP interrupt sources */
const char *iopIntNames[] = {
    "VBLANK Start",
    "GPU",
    "CDVD",
    "DMA",
    "Timer 0", "Timer 1", "Timer 2", /* IOP timer interrupts */
    "SIO0", "SIO1", /* Serial I/O interrupts */
    "SPU2",
    "PIO",
    "VBLANK End",
    "DVD",
    "PCMCIA",
    "Timer 3", "Timer 4", "Timer 5", /* IOP timer interrupts */
    "SIO2",
    "HTR0", "HTR1", "HTR2", "HTR3",
    "USB",
    "EXTR",
    "FireWire", "FDMA", /* FireWire interrupts */
};

/* --- INTC registers --- */

/* EE interrupt registers */
u16 intcMASK = 0, intcSTAT = 0;

/* IOP interrupt registers */
u32 iMASK = 0, iSTAT = 0;
bool iCTRL = false;

void checkInterruptIOP();

/* Returns INTC_MASK */
u16 readMask() {
    return intcMASK;
}

/* Returns INTC_STAT */
u16 readStat() {
    return intcSTAT;
}

/* Returns I_MASK */
u32 readMaskIOP() {
    return iMASK;
}

/* Returns I_STAT */
u32 readStatIOP() {
    return iSTAT;
}

/* Returns I_CTRL */
u32 readCtrlIOP() {
    const auto oldCTRL = iCTRL;

    /* Reading I_CTRL turns off interrupts */
    iCTRL = false;

    checkInterruptIOP();

    return oldCTRL;
}

/* Writes INTC_MASK */
void writeMask(u16 data) {
    intcMASK = (intcMASK ^ data) & 0x7FFF;
}

/* Writes INTC_STAT */
void writeStat(u16 data) {
    intcSTAT = (intcSTAT & ~data) & 0x7FFF;
}

/* Writes I_MASK */
void writeMaskIOP(u32 data) {
    iMASK = data & 0x3FFFFFF;

    checkInterruptIOP();
}

/* Writes I_STAT */
void writeStatIOP(u32 data) {
    iSTAT = iSTAT & data;

    checkInterruptIOP();
}

/* Writes I_CTRL */
void writeCtrlIOP(u32 data) {
    iCTRL = data & 1;

    checkInterruptIOP();
}

void sendInterruptIOP(IOPInterrupt i) {
    std::printf("[INTC:IOP  ] %s interrupt request\n", iopIntNames[static_cast<int>(i)]);

    iSTAT |= 1 << static_cast<int>(i);

    checkInterruptIOP();
}

void checkInterruptIOP() {
    //std::printf("[INTC:IOP  ] I_CTRL = %d, I_STAT = 0x%07X, I_MASK = 0x%07X\n", iCTRL, iSTAT, iMASK);

    iop::cop0::setInterruptPending(iCTRL && (iSTAT & iMASK));
}

}
